README.md
# 4 Clojure Solutions

A collection of solutions to [4clojure.com](http://4clojure.com) problems.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-02-25
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-02-25
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/for-clojure/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/for-clojure/compare/0.1.0...0.1.1


test/for_clojure/easy_test.clj
(ns for-clojure.easy-test
  (:require [clojure.test :refer :all]
            [for-clojure.easy :refer :all]))


(deftest easy-tests

  ;; Problem 19
  (testing "Last Element"
    (is (= (last-element [1 2 3 4 5]) 5))
    (is (= (last-element '(5 4 3)) 3))
    (is (= (last-element ["b" "c" "d"]) "d")))


  ;; Problem 20
  (testing "Penultimate Element"
    (is (= (penulimate-element (list 1 2 3 4 5)) 4))
    (is (= (penulimate-element ["a" "b" "c"]) "b"))
    (is (= (penulimate-element [[1 2] [3 4]]) [1 2])))


  ;; Problem 21
  (testing "Nth Element"
    (is (= (nth-element '(4 5 6 7) 2) 6))
    (is (= (nth-element [:a :b :c] 0) :a))
    (is (= (nth-element [1 2 3 4] 1) 2))
    (is (= (nth-element '([1 2] [3 4] [5 6]) 2) [5 6])))


  ;; Problem 22
  (testing "Count a Sequence"
    (is (= (count-a-sequence '(1 2 3 3 1)) 5))
    (is (= (count-a-sequence "Hello World") 11))
    (is (= (count-a-sequence [[1 2] [3 4] [5 6]]) 3))
    (is (= (count-a-sequence '(13)) 1))
    (is (= (count-a-sequence '(:a :b :c)) 3)))


  ;; Problem 23
  (testing "Reverse a Sequence"
    (is (= (reverse-a-sequence [1 2 3 4 5]) [5 4 3 2 1]))
    (is (= (reverse-a-sequence (sorted-set 5 7 2 7)) '(7 5 2)))
    (is (= (reverse-a-sequence [[1 2][3 4][5 6]]) [[5 6][3 4][1 2]])))


  ;; Problem 24
  (testing "Sum It All Up"
    (is (= (sum-it-all-up [1 2 3]) 6))
    (is (= (sum-it-all-up (list 0 -2 5 5)) 8))
    (is (= (sum-it-all-up #{4 2 1}) 7))
    (is (= (sum-it-all-up '(0 0 -1)) -1))
    (is (= (sum-it-all-up '(1 10 3)) 14)))


  ;; Problem 25
  (testing "Find the odd numbers"
    (is (= (find-the-odd-numbers #{1 2 3 4 5}) '(1 3 5)))
    (is (= (find-the-odd-numbers [4 2 1 6]) '(1)))
    (is (= (find-the-odd-numbers [2 2 4 6]) '()))
    (is (= (find-the-odd-numbers [1 1 1 3]) '(1 1 1 3))))


  ;; Problem 26
  (testing "Fibonacci Sequence"
    (is (= (fibonacci-sequence 3) '(1 1 2)))
    (is (= (fibonacci-sequence 6) '(1 1 2 3 5 8)))
    (is (= (fibonacci-sequence 8) '(1 1 2 3 5 8 13 21))))


  ;; Problem 27
  (testing "Palindrome Detector"
    (is (false? (palindrome-detector '(1 2 3 4 5))))
    (is (true? (palindrome-detector "racecar")))
    (is (true? (palindrome-detector [:foo :bar :foo])))
    (is (true? (palindrome-detector '(1 1 3 3 1 1))))
    (is (false? (palindrome-detector '(:a :b :c)))))


  ;; Problem 28
  (testing "Flatten a Sequence"
    (is (= (flatten-a-sequence '((1 2) 3 [4 [5 6]])) '(1 2 3 4 5 6)))
    (is (= (flatten-a-sequence ["a" ["b"] "c"]) '("a" "b" "c")))
    (is (= (flatten-a-sequence '((((:a))))) '(:a))))


  ;; Problem 29
  (testing "Get the Caps"
    (is (= (get-the-caps "HeLlO, WoRlD!") "HLOWRD"))
    (is (empty? (get-the-caps "nothing")))
    (is (= (get-the-caps "$#A(*&987Zf") "AZ")))


  ;; Problem 30
  (testing "Compress a Sequence"
    (is (= (apply str (compress-a-sequence "Leeeeeerrroyyy")) "Leroy"))
    (is (= (compress-a-sequence [1 1 2 3 3 2 2 3]) '(1 2 3 2 3)))
    (is (= (compress-a-sequence [[1 2] [1 2] [3 4] [1 2]]) '([1 2] [3 4] [1 2]))))


  ;; Problem 31
  (testing "Pack a Sequence"
    (is (= (pack-a-sequence [1 1 2 1 1 1 3 3]) '((1 1) (2) (1 1 1) (3 3))))
    (is (= (pack-a-sequence [:a :a :b :b :c]) '((:a :a) (:b :b) (:c))))
    (is (= (pack-a-sequence [[1 2] [1 2] [3 4]]) '(([1 2] [1 2]) ([3 4])))))


  ;; Problem 32
  (testing "Duplicate a Sequence"
    (is (= (duplicate-a-sequence [1 2 3]) '(1 1 2 2 3 3)))
    (is (= (duplicate-a-sequence [:a :a :b :b]) '(:a :a :a :a :b :b :b :b)))
    (is (= (duplicate-a-sequence [[1 2] [3 4]]) '([1 2] [1 2] [3 4] [3 4])))
    (is (= (duplicate-a-sequence [[1 2] [3 4]]) '([1 2] [1 2] [3 4] [3 4]))))



  ;; Problem 33
  (testing "Replicate a Sequence"
    (is (= (replicate-a-sequence [1 2 3] 2) '(1 1 2 2 3 3)))
    (is (= (replicate-a-sequence [:a :b] 4) '(:a :a :a :a :b :b :b :b)))
    (is (= (replicate-a-sequence [4 5 6] 1) '(4 5 6)))
    (is (= (replicate-a-sequence [[1 2] [3 4]] 2) '([1 2] [1 2] [3 4] [3 4])))
    (is (= (replicate-a-sequence [44 33] 2) [44 44 33 33])))


  ;; Problem 34
  (testing "Implement range"
    (is (= (implement-range 1 4) '(1 2 3)))
    (is (= (implement-range -2 2) '(-2 -1 0 1)))
    (is (= (implement-range 5 8) '(5 6 7))))
)




test/for_clojure/elementary_test.clj
(ns for-clojure.elementary-test
  (:require [clojure.test :refer :all]
            [clojure.set]
            [for-clojure.elementary :refer :all]))


(deftest elementary-tests

  ;; Problem 1
  (testing "Nothing but the Truth"
    (is (= nothing-but-the-truth true)))


  ;; Problem 2
  (testing "Simple Math"
    (is (= (- 10 (* 2 3)) simple-math)))


  ;; Problem 3
  (testing "Intro to Strings"
    (is (= intro-to-strings (.toUpperCase "hello world"))))


  ;; Problem 4
  (testing "Intro to Lists"
    (is (= intro-to-lists '(:a :b :c))))


  ;; Problem 5
  (testing "Lists: conj"
    (is (= lists-conj (conj '(2 3 4) 1)))
    (is (= lists-conj (conj '(3 4) 2 1))))


  ;; Problem 6
  (testing "Intro to Vectors"
    ;; TODO make it run like an original problem
    (is (= intro-to-vectors (list :a :b :c) (vec '(:a :b :c)) (vector :a :b :c))))


  ;; Problem 7
  (testing "Vectors: conj"
    (is (= vectors-conj (conj [1 2 3] 4)))
    (is (= vectors-conj (conj [1 2] 3 4))))


  ;; Problem 8
  (testing "Intro to Sets"
    (is (= intro-to-sets (set '(:a :a :b :c :c :c :c :d :d))))
    (is (= intro-to-sets (clojure.set/union #{:a :b :c} #{:b :c :d}))))


  ;; Problem 9
  (testing "Sets: conj"
    (is (= #{1 2 3 4} (conj #{1 4 3} sets-conj))))


  ;; Problem 10
  (testing "Intro to Maps"
    (is (= intro-to-maps ((hash-map :a 10, :b 20, :c 30) :b))))


  ;; Problem 11
  (testing "Maps: conj"
    (is (= {:a 1, :b 2, :c 3} (conj {:a 1} maps-conj [:c 3]))))


  ;; Problem 12
  (testing "Intro to Sequences"
    (is (= intro-to-sequences (first '(3 2 1))))
    (is (= intro-to-sequences (second [2 3 4])))
    (is (= intro-to-sequences (last (list 1 2 3)))))


  ;; Problem 13
  (testing "Sequences: rest"
    (is (= sequences-rest (rest [10 20 30 40]))))


  ;; Problem 14
  (testing "Intro to Functions"
    (is (= intro-to-functions ((fn add-five [x] (+ x 5)) 3)))
    (is (= intro-to-functions ((fn [x] (+ x 5)) 3)))
    (is (= intro-to-functions (#(+ % 5) 3)))
    (is (= intro-to-functions ((partial + 5) 3))))


  ;; Problem 15
  (testing "Double Down"
    (is (= (double-down 2) 4))
    (is (= (double-down 3) 6))
    (is (= (double-down 11) 22))
    (is (= (double-down 7) 14)))


  ;; Problem 16
  (testing "Hello World"
    (is (= (hello-world "Dave") "Hello, Dave!"))
    (is (= (hello-world "Jenn") "Hello, Jenn!"))
    (is (= (hello-world "Rhea") "Hello, Rhea!")))


  ;; Problem 17
  (testing "Sequences: map"
    (is (= sequences-map (map #(+ % 5) '(1 2 3)))))


  ;; Problem 18
  (testing "Sequences: filter"
    (is (= sequences-filter (filter #(> % 5) '(3 4 5 6 7)))))


  ;; Problem 35
  (testing "Local binding"
    (is (= local-binding (let [x 5] (+ 2 x))))
    (is (= local-binding (let [x 3, y 10] (- y x))))
    (is (= local-binding (let [x 21] (let [y 3] (/ x y))))))


  ;; Problem 36
  (testing "Let it Be"
    (= 10 (let [x 7 y 3 z 1] (+ x y)))
    (= 4  (let [x 7 y 3 z 1] (+ y z)))
    (= 1  (let [x 7 y 3 z 1] z)))


  ;; Problem 37
  (testing "Regular Expressions"
    (is (= regular-expressions (apply str (re-seq #"[A-Z]+" "bA1B3Ce ")))))


  ;; Problem 52
  (testing "Intro to Destructuring"
    (is (= [2 4] (let [[a b c d e] [0 1 2 3 4]] [c e]))))


  ;; Problem 57
  (testing "Simple Recursion"
    (is (= simple-recursion
           ((fn foo [x]
              (when (> x 0)
                (conj (foo (dec x)) x))) 5))))


  ;; Problem 64
  (testing "Intro to Reduce"
    (is (= 15 (reduce intro-to-reduce [1 2 3 4 5])))
    (is (=  0 (reduce intro-to-reduce [])))
    (is (=  6 (reduce intro-to-reduce 1 [2 3]))))


  ;; Problem 68
  (testing "Recurring Theme"
    (is (= recurring-theme
           (loop [x 5
                  result []]
             (if (> x 0)
               (recur (dec x) (conj result (+ 2 x)))
               result)))))


  ;; Problem 71
  (testing "Rearranging Code: ->"
    (is (= (rearranging-code-> (sort (rest (reverse [2 5 4 1 3 6]))))
           (-> [2 5 4 1 3 6] (reverse) (rest) (sort) (rearranging-code->))
           5)))


  ;; Problem 72
  (testing "Rearranging Code: ->>"
    (is (= (rearranging-code->> (map inc (take 3 (drop 2 [2 5 4 1 3 6]))))
   (->> [2 5 4 1 3 6] (drop 2) (take 3) (map inc) (rearranging-code->>))
   11)))


  ;; Problem 134
  (testing "A nil key"
    (is (true?  (a-nil-key :a {:a nil :b 2})))
    (is (false? (a-nil-key :b {:a nil :b 2})))
    (is (false? (a-nil-key :c {:a nil :b 2}))))



  ;; Problem 145
(testing "For the win"
  (is (= for-the-win (for [x (range 40)
                           :when (= 1 (rem x 4))]
                       x)))
  (is (= for-the-win (for [x (iterate #(+ 4 %) 0)
                           :let [z (inc x)]
                           :while (< z 40)]
                       z)))
  (is (= for-the-win (for [[x y] (partition 2 (range 20))]
                       (+ x y)))))


  ;; Problem 156
  (testing "Map Defaults"
    (is (= (map-defaults 0 [:a :b :c]) {:a 0 :b 0 :c 0}))
    (is (= (map-defaults "x" [1 2 3]) {1 "x" 2 "x" 3 "x"}))
    (is (= (map-defaults [:a :b] [:foo :bar]) {:foo [:a :b] :bar [:a :b]})))


  ;; Problem 161
  (testing "Subset and Superset"
    (is (clojure.set/superset? subset-and-superset #{2}))
    (is (clojure.set/subset? #{1} subset-and-superset))
    (is (clojure.set/superset? subset-and-superset #{1 2}))
    (is (clojure.set/subset? #{1 2} subset-and-superset)))


  ;; Problem 162
  (testing "Logical falsity and truth"
    (is (= logical-falsity-and-truth (if-not false 1 0)))
    (is (= logical-falsity-and-truth (if-not nil 1 0)))
    (is (= logical-falsity-and-truth (if true 1 0)))
    (is (= logical-falsity-and-truth (if [] 1 0)))
    (is (= logical-falsity-and-truth (if [0] 1 0)))
    (is (= logical-falsity-and-truth (if 0 1 0)))
    (is (= logical-falsity-and-truth (if 1 1 0))))

)




src/for_clojure/easy.clj
(ns for-clojure.easy)


;; Problem 19
(defn last-element [xs]
  (-> xs reverse first))


;; Problem 20
(defn penulimate-element [xs]
  (-> xs reverse second))


;; Problem 21
(defn nth-element [xs n]
  (first (drop n xs)))

;; Problem 22
(defn count-a-sequence [xs]
  (reduce (fn [c _] (inc c)) 0 xs))


;; Problem 23
(defn reverse-a-sequence [xs]
  (reduce #(cons %2 %1) '() xs))


;; Problem 24
(defn sum-it-all-up [xs]
  (reduce + xs))


;; Problem 25
(defn find-the-odd-numbers [xs]
  (filter odd? xs))


;; Problem 26
(defn fibonacci-sequence [n]
  (let [fib-seq ((fn fib [a b] (lazy-seq (cons b (fib b (+ a b))))) 0 1)]
    (take n fib-seq)))


;; Problem 27
(defn palindrome-detector [s]
  (= (seq s) (reverse s)))


;; Problem 28
(defn flatten-a-sequence [xs]
  (when-let [[head & tail] (seq xs)]
    (if (coll? head)
      (concat (flatten-a-sequence head)
              (flatten-a-sequence tail))
      (cons head (flatten-a-sequence tail)))))


;; Problem 29
(defn get-the-caps [s]
  (apply str (re-seq #"[A-Z]" s)))


;; Problem 30
(defn compress-a-sequence [xs]
  (when-let [[head & tail] (seq xs)]
    (if (not= head (first tail))
      (cons head (compress-a-sequence tail))
      (compress-a-sequence tail))))


;; Problem 31
(defn pack-a-sequence [xs]
  (partition-by identity xs))


;; Problem 32
(defn duplicate-a-sequence [xs]
  (when-let [[head & tail] xs]
    (conj (duplicate-a-sequence tail) head head)))


;; Problem 33
(defn replicate-a-sequence [xs n]
  (->> xs (map #(repeat n %)) (apply concat)))


;; Problem 34
(defn implement-range [a b]
  (if (= a b)
    '()
    (cons a (implement-range (inc a) b))))



src/for_clojure/elementary.clj
(ns for-clojure.elementary)


;; Problem 1
(def nothing-but-the-truth true)


;; Problem 2
(def simple-math 4)


;; Problem 3
(def intro-to-strings "HELLO WORLD")


;; Problem 4
(def intro-to-lists '(:a :b :c))


;; Problem 5
(def lists-conj '(1 2 3 4))


;; Problem 6
(def intro-to-vectors [:a :b :c])


;; Problem 7
(def vectors-conj [1 2 3 4])


;; Problem 8
(def intro-to-sets #{:a :b :c :d})


;; Problem 9
(def sets-conj 2)


;; Problem 10
(def intro-to-maps 20)


;; Problem 11
(def maps-conj [:b 2])


;; Problem 12
(def intro-to-sequences 3)


;; Problem 13
(def sequences-rest '(20 30 40))


;; Problem 14
(def intro-to-functions 8)


;; Problem 15
(defn double-down [number] (* number 2))


;; Problem 16
(defn hello-world [name] (str "Hello, " name "!"))


;; Problem 17
(def sequences-map '(6 7 8))


;; Problem 18
(def sequences-filter '(6 7))


;; Problem 35
(def local-binding 7)


;; Problem 36
;; It hardcoded in tests:
;; [x 7 y 3 z 1]


;; Problem 37
(def regular-expressions "ABC")


;; Problem 52
;; It hardcoded in tests:
;; [c e]


;; Problem 57
(def simple-recursion '(5 4 3 2 1))


;; Problem 64
(def intro-to-reduce +)


;; Problem 68
(def recurring-theme [7 6 5 4 3])


;; Problem 71
(def rearranging-code-> last)


;; Problem 72
(def rearranging-code->> #(apply + %))


;; Problem 134
(defn a-nil-key [k m] (and (contains? m k)
                           (nil? (k m))))


;; Problem 145
(def for-the-win '(1 5 9 13 17 21 25 29 33 37))


;; Problem 156
(defn map-defaults [v ks] (zipmap ks (repeat v)))


;; Problem 161
(def subset-and-superset #{1 2})


;; Problem 162
(def logical-falsity-and-truth 1)
